### PR TITLE
Enforce doubly-linked chained block

### LINKED ISSUE #329: Assertion fail when running jit-bf with ENABLE_JIT=1
Error message:
```
rv32emu: src/emulate.c:647: block_translate: Assertion `prev_ir' failed.
Aborted (core dumped)
```

The way to reproduce:
```
make clean && make ENABLE_JIT=1
./build/rv32emu ./build/jit-bf.elf
```

### PR DESCRIPTION
**Pull request by qwe661234**
Previously, the chained block was a linear structure where previous block only pointed to the next block. Now, we have introduced additional information, allowing the next block to also reference the previous block. This modification enhance the deletion of replaced block.

Close: #329

### CODE REVIEW

File: src/utils.h [hunk 1]
```diff
@@ -56,3 +56,65 @@ static inline uintptr_t align_up(uintptr_t sz, size_t alignment)
         return ((sz + mask) & ~mask);
     return (((sz + mask) / alignment) * alignment);
 }
+
+struct list_head {
+    struct list_head *prev, *next;
+};
+
+inline void INIT_LIST_HEAD(struct list_head *head)
```
Comment (jserv) [inline src/utils.h]: Use `static inline` rather than `inline`.

File: src/utils.h [hunk 2]
```diff
@@ -56,3 +56,65 @@ static inline uintptr_t align_up(uintptr_t sz, size_t alignment)
...
+{
+    head->next = head;
+    head->prev = head;
+}
+
+inline int list_empty(const struct list_head *head)
```
Comment (jserv) [inline src/utils.h]: Specify `bool` for the return type.

File: src/emulate.c [hunk 3]
```diff
@@ -1016,16 +1052,19 @@ void rv_step(riscv_t *rv, int32_t cycles)
...
+                prev = block_find(&rv->block_map, last_pc);
+#else
+                prev = cache_get(rv->block_cache, last_pc);
+#endif
         }
-
+        /* lookup the next block in block map or translate a new block,
+         * and move onto the next block.
+         */
+        block = block_find_or_translate(rv);
```
Comment (jserv) [inline src/emulate.c]: Declare `block` here.

File: src/emulate.c [hunk 4]
```diff
@@ -1016,16 +1052,19 @@ void rv_step(riscv_t *rv, int32_t cycles)
...
         block_t *block;
-        /* try to predict the next block */
-        if (prev && prev->predict && prev->predict->pc_start == rv->PC) {
-            block = prev->predict;
-        } else {
-            /* lookup the next block in block map or translate a new block,
-             * and move onto the next block.
-             */
-            block = block_find_or_translate(rv);
+        if (prev) {
```
Comment (jserv) [inline src/emulate.c]: Rewrite as following:
```c
        if (prev && prev->pc_start != last_pc) {
            /* update previous block */
#if !RV32_HAS(JIT)
            prev = block_find(&rv->block_map, last_pc);
#else
            prev = cache_get(rv->block_cache, last_pc);
#endif
        }
```

File: src/emulate.c [hunk 5]
```diff
@@ -1034,28 +1073,41 @@ void rv_step(riscv_t *rv, int32_t cycles)
...
-                           || IF_insn(last_ir, cj) || IF_insn(last_ir, cjal)
+                       || IF_insn(last_ir, cj) || IF_insn(last_ir, cjal)
 #endif
-                ) {
+            ) {
+                if (!last_ir->branch_taken) {
                     last_ir->branch_taken = block->ir_head;
+#if RV32_HAS(JIT)
+                    chaining_entry_t *new_entry =
+                        malloc(sizeof(chaining_entry_t));
```
Comment (jserv) [inline src/emulate.c]: Shall we utilize memory pool when possible?

File: src/emulate.c [hunk 6]
```diff
@@ -979,6 +979,49 @@ static block_t *block_find_or_translate(riscv_t *rv)
...
+                if (entry->block == delete_target)
+                    continue;
+                rv_insn_t *target = entry->block->ir_tail;
+                if (target->branch_taken == delete_target->ir_head)
+                    target->branch_taken = NULL;
+                else if (target->branch_untaken == delete_target->ir_head)
+                    target->branch_untaken = NULL;
+                free(entry);
+            }
+            /* free delete block */
```
Comment (jserv) [inline src/emulate.c]: Fix the grammar mistake.

File: src/utils.h [hunk 7]
```diff
@@ -56,3 +56,65 @@ static inline uintptr_t align_up(uintptr_t sz, size_t alignment)
...
     return (((sz + mask) / alignment) * alignment);
 }
+
+struct list_head {
+    struct list_head *prev, *next;
+};
+
+inline void INIT_LIST_HEAD(struct list_head *head)
+{
+    head->next = head;
```
Comment (jserv) [inline src/utils.h]: Rewrite as `head->next = head->prev = head;`

**Review by jserv — CHANGES_REQUESTED** [review]
Explicitly mention `Close #329` in the body of git commit messages.

File: src/riscv_private.h [hunk 8]
```diff
@@ -68,9 +68,17 @@ typedef struct block {
...
         translatable; /**< Determine the block has RV32AF insturctions or not */
+    struct list_head list;
 #endif
 } block_t;
 
+#if RV32_HAS(JIT)
+typedef struct chaining_entry {
+    block_t *block;
+    struct list_head list;
+} chaining_entry_t;
```
Comment (jserv) [inline src/riscv_private.h]: Simply name it `chain_entry_t`.

File: src/riscv_private.h [hunk 9]
```diff
@@ -68,9 +68,17 @@ typedef struct block {
     uint32_t offset;
     bool
         translatable; /**< Determine the block has RV32AF insturctions or not */
+    struct list_head list;
 #endif
 } block_t;
 
+#if RV32_HAS(JIT)
+typedef struct chaining_entry {
```
Comment (jserv) [inline src/riscv_private.h]: Use anonymous struct.

File: src/riscv_private.h [hunk 10]
```diff
@@ -117,6 +125,7 @@ struct riscv_internal {
     block_map_t block_map; /**< basic block map */
 #else
     struct cache *block_cache;
+    struct mpool *chaining_entry_mp;
```
Comment (jserv) [inline src/riscv_private.h]: Ditto. Use shorter form. i.e., `chain_entry_pool`.

**Review by jserv — CHANGES_REQUESTED** [review]
Be aware of the terminology. It is referred as "doubly-linked chained block" instead of "doubly-link chained block."

File: src/emulate.c [hunk 11]
```diff
@@ -979,6 +979,49 @@ static block_t *block_find_or_translate(riscv_t *rv)
         /* insert the block into block cache */
         block_t *delete_target = cache_put(rv->block_cache, rv->PC, &(*next));
         if (delete_target) {
+            if (prev == delete_target)
+                prev = NULL;
+            chain_entry_t *entry, *safe;
+            /* correctly remove delete block from its chained block */
```
Comment (jserv) [inline src/emulate.c]: should be "deleted block"

File: src/emulate.c [hunk 12]
```diff
@@ -979,6 +979,49 @@ static block_t *block_find_or_translate(riscv_t *rv)
...
+                list_for_each_entry_safe (entry, safe, &target->list, list) {
+                    if (entry->block == delete_target) {
+                        list_del_init(&entry->list);
+                        mpool_free(rv->chain_entry_mp, entry);
+                        flag = true;
+                    }
+                }
+                assert(flag);
+            }
+            /* correctly remove delete block from the block chained to it */
```
Comment (jserv) [inline src/emulate.c]: Ditto. "deleted block"

File: src/jit.c [hunk 13]
```diff
@@ -1496,14 +1496,14 @@ static void translate_chained_block(struct jit_state *state,
...
+        block_t *block1 = cache_get(rv->block_cache, ir->branch_untaken->pc);
+        if (block1->translatable)
             translate_chained_block(state, rv, block1, set);
     }
-    if (ir->branch_taken && !set_has(set, ir->pc + ir->imm)) {
-        block_t *block1 = cache_get(rv->block_cache, ir->pc + ir->imm);
-        if (block1 && block1->translatable)
+    if (ir->branch_taken && !set_has(set, ir->branch_taken->pc)) {
+        block_t *block1 = cache_get(rv->block_cache, ir->branch_taken->pc);
+        if (block1->translatable)
```
Comment (jserv) [inline src/jit.c:1506]: Shall we check `block1` before dereferencing?

Comment (qwe661234) [reply src/jit.c:1506]: The block1 must be existed because we correctly handle deleted block.

**Review by visitorckw — APPROVED** [review]

**Comment by visitorckw** [comment]
LGTM, the patches indeed addressed the issue, thanks!

**Pull request merged**